### PR TITLE
Rustify PackedFunc & Friends

### PR DESCRIPTION
**Pull request by nhynes**

### CODE REVIEW

**Comment by nhynes** [comment]
cc @ehsanmok for review

File: rust/common/src/packed_func.rs [hunk 1]
```diff
@@ -15,298 +17,305 @@ pub trait PackedFunc =
...
 }
 
-impl<'a> TVMArgValue<'a> {
-    pub fn new(value: TVMValue, type_code: i64) -> Self {
-        TVMArgValue {
-            _lifetime: PhantomData,
-            value: value,
-            type_code: type_code,
+TVMPODValue! {
+    /// A borrowed TVMPODValue. Can be constructed using `into()` but the preferred way
```
Comment (ehsanmok) [inline rust/common/src/packed_func.rs]: For docs to be generated, you need to capture them as `#[doc=$doc]` like [this](https://github.com/rust-ndarray/ndarray/blob/master/src/impl_ops.rs#L51) for example.

File: rust/common/src/packed_func.rs [hunk 2]
```diff
@@ -15,298 +17,305 @@ pub trait PackedFunc =
...
 
-/// A borrowed TVMPODValue. Can be constructed using `into()` but the preferred way
-/// to obtain a `TVMArgValue` is automatically via `call_packed!`.
-#[derive(Clone, Copy)]
-pub struct TVMArgValue<'a> {
-    pub _lifetime: PhantomData<&'a ()>,
-    pub value: TVMValue,
-    pub type_code: i64,
+/// Constructs a derivative of a TVMPodValue.
+macro_rules! TVMPODValue {
```
Comment (ehsanmok) [inline rust/common/src/packed_func.rs]: Neat :)

File: rust/frontend/src/function.rs [hunk 3]
```diff
@@ -195,11 +195,11 @@ impl<'a, 'm> Builder<'a, 'm> {
         let (mut values, mut type_codes): (Vec<ffi::TVMValue>, Vec<ffi::TVMTypeCode>) = self
             .arg_buf
             .iter()
-            .map(|tvm_arg| (tvm_arg.value, tvm_arg.type_code as ffi::TVMTypeCode))
+            .map(|arg| arg.clone().into_tvm_value())
```
Comment (ehsanmok) [inline rust/frontend/src/function.rs]: Can it be without `clone` after `into_iter` or something?

File: rust/common/src/packed_func.rs [hunk 4]
```diff
@@ -15,298 +17,305 @@ pub trait PackedFunc =
...
+    {
+        $(#[$m:meta])+
```
[the remark this answers is not in the document]
Comment (nhynes) [reply rust/common/src/packed_func.rs]: I'd considered it, but I like requiring doc comments :)

File: rust/common/src/packed_func.rs (hunk 1 above)
Comment (nhynes) [reply rust/common/src/packed_func.rs]: Rustc makes a [docs pass](https://doc.rust-lang.org/rustdoc/passes.html#passes) before expanding macros that condenses all of the doc comments into `#[doc(blah)]`

File: rust/frontend/src/function.rs (hunk 3 above)
Comment (nhynes) [reply rust/frontend/src/function.rs]: that would require

```diff
- type PackedFunc = Fn(&[TVMArgValue]) -> TVMRetValue;
+ type PackedFunc = Fn(Vec<TVMArgValue>) -> TVMRetValue;
```

and the TVM module is surely not going to know how to Drop a Vec (nor can the frontend or runtime Drop a C++ std::vector).

I think that a better solution might be `.as_tvm_value()`, but that adds a lot  of boilerplate code and it's not totally clear that rustc doesn't just optimize away the clone.

File: rust/common/src/packed_func.rs [hunk 5]
```diff
@@ -15,298 +17,305 @@ pub trait PackedFunc =
 /// `call_packed!(my_tvm_func, &mut arg1, &mut arg2)`
 #[macro_export]
 macro_rules! call_packed {
-  ($fn:expr, $($args:expr),+) => {
-    $fn(&[$($args.into(),)+])
-  };
-  ($fn:expr) => {
-    $fn(&Vec::new())
-  };
+    ($fn:expr, $($args:expr),+) => {
```
[the remark this answers is not in the document]
Comment (nhynes) [reply rust/common/src/packed_func.rs]: rustfmt doesn't format inside of macros. We updated the `.rustfmt.toml`, but it didn't apply, so I had to do it manually,

File: rust/frontend/src/function.rs (hunk 3 above)
Comment (nhynes) [reply rust/frontend/src/function.rs]: Well, `as_tvm_value` seems correct. I guess I'll just make the macros more macro-y to eliminate the repetitive code.

**Comment by ehsanmok** [comment]
@nhynes @tqchen LGTM

File: rust/frontend/src/function.rs (hunk 3 above)
Comment (kazimuth) [reply rust/frontend/src/function.rs]: You could eliminate the allocation on the rust side doing something like:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8d4048c7235d7e7bf8074f1514a07350

This wouldn't be able to go through `Builder` though.

Also, mapping a slice of tuples to a pair of slices is actually surprisingly hard to do with macros. You could avoid doing it if you're able to just call `into()` twice for each macro, or else you can do some [advanced macro magic](https://danielkeep.github.io/tlborm/book/pat-push-down-accumulation.html) to map over the slice.

**Comment by kazimuth** [comment]
You could eliminate the call_packed! allocation on the rust side doing something like:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8d4048c7235d7e7bf8074f1514a07350

This would require a rework of `Builder`, though.

**Comment by tqchen** [comment]
@kazimuth can you also https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by kazimuth** [comment]
@tqchen will do in the future, although on further inspection I think my suggestion won't actually work in this case because of the way `function::Builder` works.

**Comment by nhynes** [comment]
I replaced `into_tvm_value` with `to_tvm_value` which takes `&self`. It copies the primitives and pointers, but this is actually correct since something needs to retain ownership of the strings and bytearrays.

**Comment by ehsanmok** [comment]
@nhynes if I remember correctly, `into_tvm_value` was necessary but not sure it's relation to your changes now.

**Comment by nhynes** [comment]
It's no longer `into` because `ffi::TVMValue` doesn't own its contents. If, for instance, the frontend did `into_tvm_value`, there would be no guarantee that the pointed-to data wasn't dropped.

**Comment by ehsanmok** [comment]
@tqchen This PR is ready to be merged.

**Comment by ehsanmok** [comment]
@tqchen Could you merge this sooner please? I'm working on a feature which requires these changes.

**Comment by tqchen** [comment]
awaiting approval from @kazimuth

File: rust/frontend/src/function.rs (hunk 3 above)
Comment (kazimuth) [reply rust/frontend/src/function.rs]: This can be addressed later.

**Comment by kazimuth** [comment]
Done @tqchen

**Comment by tqchen** [comment]
THanks, @nhynes , @kazimuth @ehsanmok , this is now merged

**Pull request merged**